### PR TITLE
Add `is_sorted` to `Iterator` and `[T]`

### PR DESCRIPTION
**Pull request by kleimkuhler**
This is an initial implementation for the first step of [RFC 2351](https://github.com/rust-lang/rfcs/blob/master/text/2351-is-sorted.md)

Tracking issue: https://github.com/rust-lang/rust/issues/53485

### CODE REVIEW

**Review by LukasKalbertodt — COMMENTED** [review]
Apart from my these three nits, LGTM :+1: 
Thanks for doing this!

But I'm not from the Rust team, so this still needs an official review ^_^

File: src/libcore/slice/mod.rs
Comment (Kerollmops) [reply src/libcore/slice/mod.rs]: This way it could be interesting to use `chunk_exact` which allow llvm to do [better assumption for bounds checks](https://github.com/rust-lang/rust/issues/47115#issuecomment-354715511).

Comment (kleimkuhler) [reply src/libcore/slice/mod.rs]: Good call, thanks! So it's been changed from
```
let mut last = match self.first() {
    Some(e) => e,
    None => return true,
};

for curr in &self[1..] {
    if compare(&last, &curr).map(|o| o == Ordering::Greater).unwrap_or(true) {
        return false;
    }
    last = &curr;
}

true
```
to
```
let len = self.len();
    if len <= 1 {
        return true;
    }

for i in 1..len {
    if compare(&self[i - 1], &self[i]).map(|o| o == Ordering::Greater).unwrap_or(true) {
        return false;
    }
}

true
```

If LLVM is able to elide (almost) all bounds checks on both implementations, and one is preferred over the other, let me know!

**Comment by kleimkuhler** [comment]
If the use of `#[inline]` can also be confirmed to be correct, I have it for `is_sorted` and `is_sorted_by_key` methods, but not the `is_sorted_by` method (on both `Iterator` and `[T]`)

File: src/libcore/slice/mod.rs
Comment (LukasKalbertodt) [reply src/libcore/slice/mod.rs]: In the second example, the first `if` condition is redundant and can be removed. If `len <= 1`, the `for` loop won't run at all. That was my initial idea behind my comment -- to make the code shorter. But removing this first `if` actually makes the generated assembly worse (it generates at least one bounds check). So yes, it should stay in there, but preferably with a comment explaining why it's there.

Regarding the generated code:

- [The first version with `last` variable](https://rust.godbolt.org/z/8uBjLe)
- [The second version with indexing](https://rust.godbolt.org/z/ynJIF3)

The second version is shorter. Otherwise they look fairly identical to me. But maybe someone with a bit more knowledge of these things should comment ^_^

I prefer the second version (indexing).

Comment (Kerollmops) [reply src/libcore/slice/mod.rs]: Why don't we just use [`slice::windows`](https://doc.rust-lang.org/std/primitive.slice.html#method.windows) ?

https://rust.godbolt.org/z/4h4d42

Comment (kleimkuhler) [reply src/libcore/slice/mod.rs]: Looks like `while let` generates a little less: https://rust.godbolt.org/z/W6xDfr

Comment (LukasKalbertodt) [reply src/libcore/slice/mod.rs]: So what "matters" is not really the total number of instructions but the ones executed for each element (each loop iteration). It's 6 for [the indexing version](https://rust.godbolt.org/z/ynJIF3) and 7 for [the `windows(2)` version](https://rust.godbolt.org/z/4h4d42).

But a few remarks:
- Just counting instructions is not really a reliable way to guess the speed. So by now, we would just have to measure it to say anything about performance.
- All versions are pretty close to the optimum. And I don't expect there to be a huge difference between the versions.
- For small primitive types, we will have a SIMD implementation in the future, so we don't really need to care about that. For bigger types, the time to compare those types will be a significant part of the runtime, so again, this code doesn't matter *that* much.

So, I'd say to just take the **most idiomatic solution for now, which IMO is `for v in s.windows(2)`**. We can still build a bunch of benchmarks and improve the code later if it's necessary.

**Comment by LukasKalbertodt** [comment]
> If the use of `#[inline]` can also be confirmed to be correct, I have it for `is_sorted` and `is_sorted_by_key` methods, but not the `is_sorted_by` method (on both `Iterator` and `[T]`)

I'd be interested in that as well. So your logic is fine I guess (just inline the two "wrapper" methods). But from my C++ days I still remember that "you shall not tell the compiler when to inline and when not! The compiler is way smarter than you anyway". 

Furthermore, I think `#[inline]` in Rust is also different than one might expect -- it's by no means forcing the compiler to inline it. It just makes the function "available" when compiling other crates (as in: it will be recompiled for every crate that uses it). But I think it's not necessary here as all generic functions are available in the same sense anyway. But I might be very wrong, given that many slice methods are marked `#[inline]`.

So yes, I'd be really interested in someone explaining what's the best thing to do in this situation ^_^

File: src/libcore/slice/mod.rs
Comment (kleimkuhler) [reply src/libcore/slice/mod.rs]: So [indexing](https://rust.godbolt.org/z/ynJIF3) and [while let with `slice::windows`](https://rust.godbolt.org/z/8tGa54) both look to be good options.

At `opt-level=1` (if that is something to consider), indexing generates a little less.

At this point, like @LukasKalbertodt mentioned someone with a little more knowledge can make the call. I like how concise `while let` is, but if `opt-level=1` is what we should go by then I also lean towards indexing.

**Comment by TimNN** [comment]
Triage: Assigning a ~random person from the libs team: r? @KodrAus

File: src/libcore/slice/mod.rs
Comment (KodrAus) [reply src/libcore/slice/mod.rs]: Going with what's idiomatic initially and letting performance improve over time through targeted refactorings sounds like a good path forwards to me :+1:

File: src/libcore/iter/iterator.rs
Comment (KodrAus) [inline src/libcore/iter/iterator.rs]: We shouldn't actually need the `inline` attribute in this case, because the method is generic so is already a candidate for inlining in the target crate.

File: src/libcore/slice/mod.rs
Comment (KodrAus) [inline src/libcore/slice/mod.rs]: We shouldn't actually need the `inline` attribute in this case, because the method is generic so is already a candidate for inlining in the target crate.

**Review by KodrAus — COMMENTED** [review]
Thanks @kleimkuhler! The updates to the unstable book and the docs on these new methods are really nice :+1:

The `#[inline]` attribute is a bit of a funny one. Like @LukasKalbertodt says it doesn't force inlining, but makes inlining possible across crates if it wasn't already. My understanding is that it's only needed when you want something to be able to be inlined and there are no generics involved. If there are generics then we don't need the attribute, because of the way monomorphization works those methods are already candidates for inlining because they're compiled into the target crate.

**Comment by kleimkuhler** [comment]
Thanks for the explanation on #[inline]! I removed it from the generic methods. That currently leaves all comments addressed/resolved.

**Review by gnzlbg — CHANGES_REQUESTED** [review]
Major things to resolve:

* `slice.iter().is_sorted_by` should dispatch to `[T]::is_sorted_by` instead of `Iterator::is_sorted_by`
* not enough tests to detect regressions:
    * for user-defined types that are only `PartialOrd` (like @LukasKalbertodt set example)
    * floating-point numbers (wanna probably cover all corner cases with `NaN`, infinity, and denormals)
     * types implementing `Ord` 

---

Minor nitpicks:

* lack of `#[inline]` attribute prevents monomorphizations of these methods generated in `libcore`/`libstd` from being inlined into user crates (e.g. if `#[inline] fn libstd::foo(x: &[u32]) { x.is_sorted() }` is inlined into a user's crate, `is_sorted` itself cannot be inlined if its not `#[inline]` as well because the monomorphization happened in `libstd` instead of the user crate since the type of `x` is concrete).

File: src/libcore/tests/iter.rs
Comment (kleimkuhler) [reply src/libcore/tests/iter.rs]: Thanks for pointing this out; I was not aware that `.iter()` was returning an `Iter<T>` instead of an `Iterator<T>`. Is changing these to `Vec` (e.g. `vec!["c", "bb", "aaa"].iter()`) enough to make sure it dispatches to the implementations of the `Iterator` trait?

File: src/libcore/iter/iterator.rs
Comment (kleimkuhler) [reply src/libcore/iter/iterator.rs]: Does the same reasoning apply for `[T]`'s implementation of `is_sorted_by`? I can change both back, but I also originally had both as `#[inline]` so I just want to be sure before making the change again.

File: src/libcore/tests/iter.rs
Comment (kleimkuhler) [reply src/libcore/tests/iter.rs]: This was a test case taken from the RFC for a heads up.

File: src/libcore/iter/iterator.rs
Comment (KodrAus) [reply src/libcore/iter/iterator.rs]: Yeh, I think this one was my bad, I steered you off course. I hadn't realised that when an `#[inline]`'ed function internally calls a generic, but not explicitly `#[inline]`'ed, function then that inner function won't be inlined at all.

File: src/libcore/tests/iter.rs
Comment (LukasKalbertodt) [reply src/libcore/tests/iter.rs]: The `Iterator` impl for `slice::Iter` is generated with [this macro](https://github.com/rust-lang/rust/blob/52cc6fde4b73a3013ce2185052f9492ae3728367/src/libcore/slice/mod.rs#L2768). The macro is invoked for `slice::Iter` [here](https://github.com/rust-lang/rust/blob/52cc6fde4b73a3013ce2185052f9492ae3728367/src/libcore/slice/mod.rs#L3083) and for `slice::IterMut` [here](https://github.com/rust-lang/rust/blob/52cc6fde4b73a3013ce2185052f9492ae3728367/src/libcore/slice/mod.rs#L3184). I think we can override `is_sorted` in both impls (`Iter` and `IterMut`). So the method can just be added in the `iterator` macro.

File: src/libcore/iter/iterator.rs
Comment (gnzlbg) [reply src/libcore/iter/iterator.rs]: Arguably, the case in which this won't be inlined is contrived, and these are currently not used anywhere inside `core` yet, so that case cannot happen until then. Also, I don't think the auto-vectorizer can currently handle these methods due to the early return, so whether these are inlined or not might not be that big of a deal. 

However, if someone decides to use these somewhere inside core, chances are that they won't think about any of this, and `#[inline]` doesn't really make anything any worse since they are already generic, so I'd rather just make them `#[inline]` now.

File: src/libcore/tests/lib.rs
Comment (malbarbo) [inline src/libcore/tests/lib.rs]: Bad merge.

**Comment by TimNN** [comment]
Ping from triage @KodrAus / @rust-lang/libs: This PR needs your review.

**Comment by KodrAus** [comment]
@kleimkuhler thanks for your patience! I think we've addressed the comments about `#[inline]` now. So we just need to extend the test suite a bit with some of the curlier cases that @gnzlbg described and I think we should be good to go!

**Comment by kleimkuhler** [comment]
@KodrAus Yes sorry for the delay in this; I will address the dispatch issue and add some additional tests shortly.

I can close the PR temporarily so that it gets cleared from the triage's monitor and reopen when it is ready again if you would like!

**Comment by kleimkuhler** [comment]
I'm working on the issue of

> slice.iter().is_sorted_by should dispatch to [T]::is_sorted_by instead of Iterator::is_sorted_by

pointed out by @gnzlbg . After some input from @LukasKalbertodt I had:

```
#[inline]
fn is_sorted_by<F>(mut self, compare: F) -> bool
where
    Self: Sized,
    F: FnMut(&Self::Item, &Self::Item) -> Option<Ordering>
{
    self.make_slice().is_sorted_by(compare)
}
```

and I was getting the following error:

```
error[E0277]: expected a `ops::function::FnMut<(&T, &T)>` closure, found `F`23: core
    --> src/libcore/slice/mod.rs:3048:35
     |
3048 |                 self.make_slice().is_sorted_by(compare)
     |                                   ^^^^^^^^^^^^ expected an `FnMut<(&T, &T)>` closure, found `F`
...
3187 | iterator!{struct Iter -> *const T, &'a T, const, /* no mut */}
     | -------------------------------------------------------------- in this macro invocation
     |
     = help: the trait `for<'r, 's> ops::function::FnMut<(&'r T, &'s T)>` is not implemented for `F`
     = help: consider adding a `where for<'r, 's> F: ops::function::FnMut<(&'r T, &'s T)>` bound
```

Even with adding the lifetime bounds as recommended and getting `for<'r, 's> F: FnMut(&'r Self::Item, &'s Self::Item) -> Option<Ordering>`, I keep getting the error that the trait is not satisfied for `F`.

I went ahead with adding the same lifetime bounds to the `fn is_sorted_by`, as well as played around with the use of references and lifetimes in different places. I tried reproducing a more minimal example, but with the fact that these errors occur during the macro invocation on `#[inline]` methods I have not been too successful with that.

I'm not sure I want to keep this PR open much longer since I feel like that is just a bad reflection, but I'm not sure on my next step to address the dispatch issue. Would I be able to get some pointers on what the compiler expects with regards to the lifetime's of the references in the dispatch?

**Comment by gnzlbg** [comment]
That error message rings a bell: in `is_sorted` I solved it [this way back then](https://github.com/gnzlbg/is_sorted/blob/master/src/lib.rs#L354), but one might need to do something different here. 

My recommendation is to try to minimize the error in the playground, and try to solve it there first. People on IRC / Discord were very helpful when I was implementing this, so once you have it on the playground, you might want to ask for help there.

**Comment by LukasKalbertodt** [comment]
I reproduced the problem: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=05920a2b770a1c2e454f730aeb52dc19). The error is in fact that the closure passed to `<slice::Iter as Iterator>::is_sorted_by` receives arguments of type `&&T` (`&Self::Item` where `Self::Item` is `&T`). Sadly, we can't just write:

```rust
self.make_slice().is_sorted_by(|a, b| compare(&a, &b))
```

Because `[T]::is_sorted_by` takes `F: FnMut(&T, &T) -> Option<Ordering>`, meaning that the references the closure is called with can have any lifetime, no particularly `'a`. But since `Iterator` requires `F: FnMut(&Self::Item, &Self::Item)` which in our case means `FnMut(&&'a T, &&'a T)` (meaning: requiring exactly lifetime `'a`), we get a lifetime mismatch. In other words: while the closure of `[T]::is_sorted_by` has to be able to deal with any lifetime, the closure passed to `Iterator::is_sorted_by` can only deal with the lifetime `'a`, so we can't implement the first with the second. This is also why I think @gnzlbg solution won't work in this situation, but I am not entirely sure!

So I see two solutions:

1. Change the interface of `[T]::is_sorted_by` to:
    ```rust
    fn is_sorted_by<'a, F>(&'a self, mut compare: F) -> bool
    where
        F: FnMut(&'a T, &'a T) -> Option<Ordering>
    ```

    This "relaxes" the API as in: the caller has to provide less. But now the calle (`is_sorted_by`) really has to only pass `'a` references into `compare`. Which is ok now, but I'm not sure it will always be OK.

2. Create a private function/method with the signature shown above while keeping the public interface as specified in the RFC. Then both, `[T]::is_sorted_by` and `<slice::iter as Iterator>::is_sorted_by` would use that internal function.

I guess (2.) is the better solution here.

---

But as a general note: I'd love to see this PR get merged soon. In particular, things like "override `Iterator::is_sorted_by` for `slice::Iter`" can very well be done in subsequent PRs and shouldn't really be blocking this one. Sticking with a PR for a long time can be exhausting, in particular as a new contributor ([I know what I'm talking about](https://github.com/rust-lang/rust/pull/44042) :P). So if the overwrite-iterator thingy can't be solved easily now, I would just drop that change for this PR.

@gnzlbg Are there any things that in your opinion absolutely need to be done before merging this PR?

**Comment by gnzlbg** [comment]
> Are there any things that in your opinion absolutely need to be done before merging this PR?

I don't think the "override" has to happen in this PR, but I do think that we should choose an implementation strategy in this PR that does not require a full rewrite to allow that. Once one has the right strategy, implementing the override is a minimal change, and a "sanity" check that the strategy is correct.

> I guess (2.) is the better solution here.

Sounds good, that's pretty much what I ended up doing. In [`is_sorted/src/lib.rs`](https://github.com/gnzlbg/is_sorted/blob/master/src/lib.rs) the algorithms are all implemented in normal private `fn` (e.g. [`fn is_sorted_by_scalar_impl(...)`](https://github.com/gnzlbg/is_sorted/blob/master/src/lib.rs#L194), [`fn is_sorted_by_scalar_slice_impl(...)`](https://github.com/gnzlbg/is_sorted/blob/master/src/lib.rs#L219), etc.), which the iterator and slice implementations just call. The specialization of `SliceIter` just calls the slice function instead.

**Comment by kleimkuhler** [comment]
@LukasKalbertodt Thank you for the explanation and playground example. Unfortunately, in trying to follow along with the lifetime issue I have not been able to reproduce a working example with the strict lifetime parameters. Adjusting your playground above, it seems like you are saying the following should [work](
https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=a89c51fb37d7176be98d1fbc79f41ec7), however it is still the same issue.

**Comment by LukasKalbertodt** [comment]
@kleimkuhler You can see a working example [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=65f56ee33c381a94e8c1bf798bef17a2). I created an internal function (as suggested in my last comment, possibility 2). I also wrote a few comments. I hope this helps you, getting it to work in the real code base. If you have questions, just ask here or via Discord ^_^

**Comment by kleimkuhler** [comment]
Note: I'm unable to get a perfect replica of the error I currently have in the below playground, but I can get a superset of it. As soon as I define an `is_sorted_by_mut` trait method (which I am not doing in the actual implementation, it's just for reproducing the error here), the last two errors change, so just ignore the first error saying that the method is not part of the trait.

In this [playground](
https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=23a8170491348d77e947a12fb0060e59), the method `is_sorted_by_mut` has the type requirement that `fn compare` expects `&&'a mut T`, but it finds `&&'a T`. This is because of the return type of `fn make_slice`.

The way that I am currently trying to implement this, `Iter` and `IterMut` both dispatch to the `iterator!` macro which uses `fn make_slice`. This is an issue, because `Self::Item` expands to `&'a T` for `Iter` and `&'a mut T` for `IterMut`. This changes the expected type of `compare` in the type bound.

I think a possible way forward is to move the `is_sorted_by` method into the separate `impl`s of `Iter` and `IterMut`, but because it needs to be part of the `impl Iterator`, we would then need to move the entire `impl Iterator` into each specific struct.

I also tried extending the `iterator!` macro to take a `$container` parameter that `Iter` can pass in `&'a [T]` and `IterMut` can pass in `&'a mut [T]`, but was still limited by the `from_raw_parts` in `fn make_slice` (which would need to be `from_raw_parts_mut` for `IterMut`).

I'm not sure if I'm on the right path with this and am currently in a loop of making little changes and compiling; any recommendations?

Also - not sure why the build is failing now on the feature gates. Have not had a chance to see why they would have changed, possibly something in a rebase when addressing the merge conflicts that are popping up. This whole PR is pretty jumbled at this point so I may just loop back and make a fresh one when I can figure out a path forward. Sorry about all this.

**Comment by LukasKalbertodt** [comment]
I took another look at this problem. While it seems trivial, it's really far more annoying and problematic. We all know that Rust's capabilities of abstracting over mutability are very bad. While I thought I would solve this problem in 30 minutes or so, I am several hours in. And that's just ridiculous. Several points:

- **Without dynamic dispatch, the compiler optimizes almost perfectly**: please see [the assembly output](https://rust.godbolt.org/z/egXhq4) comparing the manual slice implementation and the generic iterator version that is called with `slice::Iter`. The assembly is not identical, but no version looks worse than the other. In fact, the generic version has one instructions fewer in the loop body. So overriding the iterator method here is not necessary at all.

- **Calling `is_sorted*` on `slice::IterMut` is exceedingly rare**: no one is going to write `slice.iter_mut().is_sorted()`. Yes, `<IterMut as Iterator>::is_sorted*` could be called in generic contexts, but I also think this is very rare. Note that `Iterator::is_sorted*` consumes `self`, so you can't do anything else with that iterator later.

- **`Iterator` in dynamic context is very rare**: the only speed benefit we can get from overriding the iterator method is for dynamic dispatch situations. And I doubt these happen often.

- **Implementing `<IterMut as Iterator>::is_sorted_by` is very tricky and will look very close to the generic implementation**: since we can't just `let a = &mut s[0]; let b = &mut s[1];` (borrowed mutably more than once), we have to use `[T]::split_first_mut` and similar methods. And that's not even all. Go ahead, [try implementing it](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=69a1830d0f2863166af61736f0c479c5). 

With that said, I wrote the code to override the method for `Iter`: see https://github.com/kleimkuhler/rust/pull/1

IMO we should merge this PR either with my additional commit or without overriding iterator methods at all. This is only useful in rare cases and shouldn't postpone this PR any longer. If anyone needs this, a future PR can override the methods. My commit at least shows that it's possible.

What do you think @gnzlbg?

**Comment by clarcharr** [comment]
Once this is rebased I'll rebase #56932 on top of this as well, as I think that it's better to merge this first.

**Comment by KodrAus** [comment]
I'll give this the ol' r+ after a rebase. The conflicts are localized to just a few `#![feature]` statements.

**Comment by kleimkuhler** [comment]
@clarcharr Thanks for the ping and sorry about the delay in getting back to you! Just returned home from some travel abroad and catching up on notifications. If you are okay to wait until end of tomorrow, I'll take some time to rebase this and get the checks passing again for a merge.

**Comment by clarcharr** [comment]
@kleimkuhler sounds good to me! I'll rebase on top of your work once it's done. I figured I'd check on the remaining Iterator PRs and make the offer. :)

**Comment by KodrAus** [comment]
Thanks for working through this @kleimkuhler!

I'll capture the outstanding questions in the tracking issue that can be followed up in future PRs so we can keep the ball rolling.

@bors r+

**Comment by timvermeulen** [comment]
The RFC points out that `Iterator::is_sorted_by_key` can be implemented using `map` + `is_sorted`, yet here it's implemented as
```rust
self.is_sorted_by(|a, b| f(a).partial_cmp(&f(b)))
```
It seems to me like `f` is now called twice for all elements (apart from the first and last ones), unlike the `map` + `is_sorted` approach. Besides, this implementation's closure has the bound `FnMut(&Self::Item) -> K` instead of the more useful `FnMut(Self::Item) -> K`.

Is there any reason why the current implementation would be preferred?

**Pull request merged**
